File: src/typing/component_declaration_config.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast
module Flow = Flow_js
open Reason

(* This is an implementation for the component declaration param handling. Given params
 * from the component_sig.DeclarationParamConfig representation, we turn those params into
 * types and define a function for extracting the config type of the component *)
module Make (Destructuring : Destructuring_sig.S) (Statement : Statement_sig.S) :
  Component_params_intf.Config with module Types = Component_sig_types.DeclarationParamConfig =
struct
  module Types = Component_sig_types.DeclarationParamConfig
  open Types

  let read_react cx loc =
    match Context.react_runtime cx with
    | Options.ReactRuntimeClassic ->
      let (_ : Type.t) =
        Type_env.var_ref ~lookup_mode:Type_env.LookupMode.ForValue cx (OrdinaryName "React") loc
      in
      ()
    | Options.ReactRuntimeAutomatic -> ()

  let param_type_with_name (Param { t; name; default; pattern; _ }) =
    let open Ast.Statement.ComponentDeclaration.Param in
    let t =
      match pattern with
      | Id id ->
        let { Ast.Pattern.Identifier.optional; _ } = id in
        let t =
          if optional || default <> None then
            TypeUtil.optional t
          else
            t
        in
        t
      | _ ->
        if default <> None then
          TypeUtil.optional t
        else
          t
    in
    match name with
    | Identifier (loc, { Ast.Identifier.name; _ }) -> (loc, name, t)
    | StringLiteral (loc, { Ast.StringLiteral.value; _ }) -> (loc, value, t)

  let rest_type (Rest { t; pattern; _ }) =
    let optional =
      match pattern with
      | Id { Ast.Pattern.Identifier.optional; _ } -> optional
      | _ -> false
    in
    if optional then
      TypeUtil.optional t
    else
      t

  let destruct cx ~use_op ~name_loc name default t =
    Base.Option.iter
      ~f:(fun d ->
        let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
        let default_t = Flow.mk_default cx reason d in
        Flow.flow cx (default_t, Type.UseT (use_op, t)))
      default;
    t

  let flow_default cx annot_t default_t =
    let open Type in
    let use_op =
      Op
        (AssignVar
           { var = Some (TypeUtil.reason_of_t annot_t); init = TypeUtil.reason_of_t default_t }
        )
    in
    Flow.flow cx (default_t, UseT (use_op, annot_t))

  let eval_default cx ?(always_flow_default = false) annot_t has_anno =
    Base.Option.map ~f:(fun e ->
        let (((loc, default_t), expr) as e) = Statement.expression cx e in
        if has_anno then (
          flow_default cx annot_t default_t;
          ((loc, annot_t), expr)
        ) else (
          if always_flow_default then flow_default cx annot_t default_t;
          e
        )
    )

  let eval_param cx (Param { t; loc; ploc; pattern; default; has_anno; name = prop_name; shorthand })
      =
    let reconstruct_prop_name t =
      let open Ast.Statement.ComponentDeclaration.Param in
      match prop_name with
      | Identifier (loc, name) -> Identifier ((loc, t), name)
      | StringLiteral lit -> StringLiteral lit
    in

    match pattern with
    | Id ({ Ast.Pattern.Identifier.name = ((name_loc, _), { Ast.Identifier.name; _ }); _ } as id) ->
      let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
      let t = Type_env.find_write cx Env_api.OrdinaryNameLoc reason in
      let default = eval_default cx ~always_flow_default:true t has_anno default in
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.Param.local = ((ploc, t), Ast.Pattern.Identifier id);
          default;
          shorthand;
          name = reconstruct_prop_name t;
        }
      )
    | Object { annot; properties; comments } ->
      let default = eval_default cx t has_anno default in
      let properties =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.object_properties cx ~f ~parent_loc:ploc init properties
      in
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.Param.local =
            ((ploc, t), Ast.Pattern.Object { Ast.Pattern.Object.properties; annot; comments });
          default;
          shorthand;
          name = reconstruct_prop_name t;
        }
      )
    | Array { annot; elements; comments } ->
      let default = eval_default cx t has_anno default in
      let elements =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.array_elements cx ~f init elements
      in
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.Param.local =
            ((ploc, t), Ast.Pattern.Array { Ast.Pattern.Array.elements; annot; comments });
          default;
          shorthand;
          name = reconstruct_prop_name t;
        }
      )

  let eval_rest cx (Rest { t; loc; ploc; pattern; has_anno }) =
    match pattern with
    | Id id ->
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.RestParam.argument =
            ((ploc, t), Ast.Pattern.Identifier id);
          comments = None;
        }
      )
    | Object { annot; properties; comments } ->
      let properties =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.object_properties cx ~f ~parent_loc:ploc init properties
      in
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.RestParam.argument =
            ((ploc, t), Ast.Pattern.Object { Ast.Pattern.Object.properties; annot; comments });
          comments = None;
        }
      )
    | Array { annot; elements; comments } ->
      let elements =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.array_elements cx ~f init elements
      in
      ( loc,
        {
          Ast.Statement.ComponentDeclaration.RestParam.argument =
            ((ploc, t), Ast.Pattern.Array { Ast.Pattern.Array.elements; annot; comments });
          comments = None;
        }
      )
end

File: rust_port/crates/flow_typing_statement/src/component_declaration_config.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_common::reason::Name;
use flow_common::reason::VirtualReasonDesc;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast;
use flow_parser::ast::statement::component_params::ParamName;
use flow_typing_context::Context;
use flow_typing_flow_js::flow_js;
use flow_typing_loc_env::component_sig_types::declaration_param_config::Param;
use flow_typing_loc_env::component_sig_types::declaration_param_config::ParamAst;
use flow_typing_loc_env::component_sig_types::declaration_param_config::Pattern;
use flow_typing_loc_env::component_sig_types::declaration_param_config::Rest;
use flow_typing_loc_env::component_sig_types::declaration_param_config::RestAst;
use flow_typing_type::type_::Type;
use flow_typing_type::type_::UseOp;
use flow_typing_type::type_util;
use flow_typing_utils::abnormal::CheckExprError;
use flow_typing_utils::type_env;

use crate::destructuring;
use crate::statement;

pub fn read_react<'a>(
    cx: &Context<'a>,
    loc: ALoc,
) -> Result<(), flow_utils_concurrency::job_error::JobError> {
    type_env::query_var(
        Some(type_env::LookupMode::ForValue),
        cx,
        Name::new(FlowSmolStr::new_inline("React")),
        None,
        loc,
    )?;
    Ok(())
}

pub fn param_type_with_name(param: &Param) -> (ALoc, FlowSmolStr, Type) {
    let t = match &param.pattern {
        Pattern::Id(id) => {
            let optional = id.optional;
            if optional || param.default.is_some() {
                type_util::optional(param.t.dupe(), None, false)
            } else {
                param.t.dupe()
            }
        }
        Pattern::Object { optional, .. } | Pattern::Array { optional, .. } => {
            if *optional || param.default.is_some() {
                type_util::optional(param.t.dupe(), None, false)
            } else {
                param.t.dupe()
            }
        }
    };
    match &param.name {
        ParamName::Identifier(id) => (id.loc.dupe(), id.name.dupe(), t),
        ParamName::StringLiteral((loc, lit)) => (loc.dupe(), lit.value.dupe(), t),
    }
}

pub fn rest_type(rest: &Rest) -> Type {
    let optional = match &rest.pattern {
        Pattern::Id(id) => id.optional,
        Pattern::Object { optional, .. } | Pattern::Array { optional, .. } => *optional,
    };
    if optional {
        type_util::optional(rest.t.dupe(), None, false)
    } else {
        rest.t.dupe()
    }
}

fn destruct<'a>(
    cx: &Context<'a>,
    use_op: &UseOp,
    name_loc: ALoc,
    name: &str,
    default: Option<&flow_typing_default::Default<Type>>,
    t: Type,
) -> Result<Type, flow_utils_concurrency::job_error::JobError> {
    if let Some(d) = default {
        let reason = flow_common::reason::mk_reason(
            VirtualReasonDesc::RIdentifier(Name::new(FlowSmolStr::new(name))),
            name_loc,
        );
        let default_t = flow_js::mk_default_non_speculating(cx, &reason, d)?;
        flow_js::flow_non_speculating(
            cx,
            (
                &default_t,
                &flow_typing_type::type_::UseT::new(flow_typing_type::type_::UseTInner::UseT(
                    use_op.dupe(),
                    t.dupe(),
                )),
            ),
        )?;
    }
    Ok(t)
}

fn flow_default<'a>(
    cx: &Context<'a>,
    annot_t: &Type,
    default_t: &Type,
) -> Result<(), flow_utils_concurrency::job_error::JobError> {
    let use_op = UseOp::Op(std::sync::Arc::new(
        flow_typing_type::type_::RootUseOp::AssignVar {
            var: Some(type_util::reason_of_t(annot_t).dupe()),
            init: type_util::reason_of_t(default_t).dupe(),
        },
    ));
    flow_js::flow_non_speculating(
        cx,
        (
            default_t,
            &flow_typing_type::type_::UseT::new(flow_typing_type::type_::UseTInner::UseT(
                use_op,
                annot_t.dupe(),
            )),
        ),
    )?;
    Ok(())
}

fn eval_default<'a>(
    cx: &Context<'a>,
    always_flow_default: bool,
    annot_t: &Type,
    has_anno: bool,
    default: Option<ast::expression::Expression<ALoc, ALoc>>,
) -> Result<Option<ast::expression::Expression<ALoc, (ALoc, Type)>>, CheckExprError> {
    default
        .map(|e| {
            let typed_e = statement::expression(None, None, None, cx, &e)?;
            let (loc, default_t) = typed_e.loc().dupe();
            if has_anno {
                flow_default(cx, annot_t, &default_t)?;
                let mut inner = (*typed_e.0).clone();
                *inner.loc_mut() = (loc, annot_t.dupe());
                Ok(ast::expression::Expression(inner.into()))
            } else {
                if always_flow_default {
                    flow_default(cx, annot_t, &default_t)?;
                }
                Ok(typed_e)
            }
        })
        .transpose()
}

pub fn eval_param<'a>(
    cx: &Context<'a>,
    param: &Param,
) -> Result<ParamAst<(ALoc, Type)>, CheckExprError> {
    let Param {
        ref t,
        ref loc,
        ref ploc,
        ref pattern,
        ref default,
        has_anno,
        shorthand,
        ref name,
    } = *param;
    let reconstruct_prop_name =
        |t: Type| -> ast::statement::component_params::ParamName<ALoc, (ALoc, Type)> {
            use ast::statement::component_params::ParamName;
            match name {
                ParamName::Identifier(id) => {
                    ParamName::Identifier(ast::Identifier::new(ast::IdentifierInner {
                        loc: (id.loc.dupe(), t),
                        name: id.name.dupe(),
                        comments: id.comments.dupe(),
                    }))
                }
                ParamName::StringLiteral(lit) => ParamName::StringLiteral(lit.clone()),
            }
        };
    match pattern {
        Pattern::Id(id) => {
            let id_name = &id.name.name;
            let name_loc = id.name.loc.0.dupe();
            let reason = flow_common::reason::mk_reason(
                VirtualReasonDesc::RIdentifier(Name::new(id_name.dupe())),
                name_loc.dupe(),
            );
            let id_t = type_env::find_write(
                cx,
                flow_env_builder::env_api::DefLocType::OrdinaryNameLoc,
                reason,
            );
            let default = eval_default(cx, true, &id_t, has_anno, default.clone())?;
            Ok(ast::statement::component_params::Param {
                loc: (loc.dupe(), id_t.dupe()),
                local: ast::pattern::Pattern::Identifier {
                    loc: (ploc.dupe(), id_t.dupe()),
                    inner: id.clone().into(),
                },
                default,
                shorthand,
                name: reconstruct_prop_name(id_t),
            })
        }
        Pattern::Object {
            annot,
            properties,
            optional,
            comments,
        } => {
            let default = eval_default(cx, false, t, has_anno, default.clone())?;
            let mut init = destructuring::empty(None, None);
            let typed_properties = destructuring::object_properties(
                cx,
                &|use_op, name_loc, name, default, t| {
                    destruct(cx, use_op, name_loc, name, default, t)
                },
                ploc.dupe(),
                &mut init,
                properties,
            )?;
            let t = t.dupe();
            Ok(ast::statement::component_params::Param {
                loc: (loc.dupe(), t.dupe()),
                local: ast::pattern::Pattern::Object {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: ast::pattern::Object {
                        properties: typed_properties.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: comments.clone(),
                    }
                    .into(),
                },
                default,
                shorthand,
                name: reconstruct_prop_name(t),
            })
        }
        Pattern::Array {
            annot,
            elements,
            optional,
            comments,
        } => {
            let default = eval_default(cx, false, t, has_anno, default.clone())?;
            let mut init = destructuring::empty(None, None);
            let typed_elements = destructuring::array_elements(
                cx,
                &|use_op, name_loc, name, default, t| {
                    destruct(cx, use_op, name_loc, name, default, t)
                },
                &mut init,
                elements,
            )?;
            let t = t.dupe();
            Ok(ast::statement::component_params::Param {
                loc: (loc.dupe(), t.dupe()),
                local: ast::pattern::Pattern::Array {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: ast::pattern::Array {
                        elements: typed_elements.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: comments.clone(),
                    }
                    .into(),
                },
                default,
                shorthand,
                name: reconstruct_prop_name(t),
            })
        }
    }
}

pub fn eval_rest<'a>(
    cx: &Context<'a>,
    rest: &Rest,
) -> Result<RestAst<(ALoc, Type)>, CheckExprError> {
    let Rest {
        ref t,
        ref loc,
        ref ploc,
        ref pattern,
        has_anno: _,
        ref comments,
    } = *rest;
    match pattern {
        Pattern::Id(id) => Ok(ast::statement::component_params::RestParam {
            loc: (loc.dupe(), t.dupe()),
            argument: ast::pattern::Pattern::Identifier {
                loc: (ploc.dupe(), t.dupe()),
                inner: id.clone().into(),
            },
            comments: comments.clone(),
        }),
        Pattern::Object {
            annot,
            properties,
            optional,
            comments: obj_comments,
        } => {
            let mut init = destructuring::empty(None, None);
            let typed_properties = destructuring::object_properties(
                cx,
                &|use_op, name_loc, name, default, dest_t| {
                    destruct(cx, use_op, name_loc, name, default, dest_t)
                },
                ploc.dupe(),
                &mut init,
                properties,
            )?;
            Ok(ast::statement::component_params::RestParam {
                loc: (loc.dupe(), t.dupe()),
                argument: ast::pattern::Pattern::Object {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: ast::pattern::Object {
                        properties: typed_properties.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: obj_comments.clone(),
                    }
                    .into(),
                },
                comments: comments.clone(),
            })
        }
        Pattern::Array {
            annot,
            elements,
            optional,
            comments: arr_comments,
        } => {
            let mut init = destructuring::empty(None, None);
            let typed_elements = destructuring::array_elements(
                cx,
                &|use_op, name_loc, name, default, dest_t| {
                    destruct(cx, use_op, name_loc, name, default, dest_t)
                },
                &mut init,
                elements,
            )?;
            Ok(ast::statement::component_params::RestParam {
                loc: (loc.dupe(), t.dupe()),
                argument: ast::pattern::Pattern::Array {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: ast::pattern::Array {
                        elements: typed_elements.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: arr_comments.clone(),
                    }
                    .into(),
                },
                comments: comments.clone(),
            })
        }
    }
}
